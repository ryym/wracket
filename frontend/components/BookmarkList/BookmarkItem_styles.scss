@import '@material/icon-button/mixins';
@import '../../styles/vars/responsive';

.itemThumbnail {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-right: var(--pad4);
  flex: 0 0 92px;
  height: 92px;
  width: 92px;

  @include screen-sm {
    flex: 0 0 68px;
    height: 68px;
    width: 68px;
  }

  > img {
    border-radius: var(--pad);
    max-width: 100%;
    max-height: 100%;
  }
}

.itemBlankThumbnail {
  background-color: rgba(#d3e6ee, 0.3);
  border-radius: var(--pad);
  width: 88%;
  height: 88%;
}

.itemLink {
  color: var(--mdc-theme-text-primary-on-background);
  text-decoration: none;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  width: 100%;
}

.itemLinkText {
  composes: mdc-list-item__text from global;
  font-size: 1.6rem;
  margin: var(--pad2) 0;
}

.itemTitle {
  line-height: 1.4em;
  display: inline-block;
  padding: var(--pad2) 0;
  white-space: initial;
}

.itemDomain {
  composes: mdc-list-item__secondary-text from global;
  font-size: 1.4rem;
  margin-right: var(--pad4);
}

.actions {
  position: absolute;
  top: 0;
  right: var(--pad);

  @include screen-sm {
    position: initial;
    width: 100%;
    text-align: right;
    margin-right: 32px;
    transition: height 0.2s;
    height: 0;
    overflow: hidden;

    &.isOpen {
      display: block;
      height: 32px;
    }
  }
}

.action {
  @include mdc-icon-button-size(18px, 18px, 4px);

  color: #b0bec5;

  @include screen-md-or-lg {
    &:hover {
      color: var(--cl-primary-900);
    }
  }

  // Remove mdc-icon-button effects on ':active'.
  &::before,
  &::after {
    opacity: 0;
    background-color: transparent;
  }

  &.isFavorite {
    color: var(--cl-secondary-400);
  }

  @include screen-sm {
    @include mdc-icon-button-size(24px, 24px, 8px);
  }
}

.actionsOpener {
  display: none;

  @include screen-sm {
    position: absolute;
    bottom: var(--pad2);
    right: var(--pad2);
    display: block;
  }

  > button {
    color: var(--cl-primary-400);

    @include mdc-icon-button-size(20px, 20px, 4px);
  }
}
