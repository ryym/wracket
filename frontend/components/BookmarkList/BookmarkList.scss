.list {
  composes: mdc-list mdc-list--two-line from global;

  .listItem {
    // Override default height 72px of '.mdc-list--two-line .mdc-list-item'.
    height: auto;
  }
}

.listItem {
  composes: mdc-list-item from global;
  border-bottom: 1px solid var(--cl-primary-50);
  min-height: 72px;
  position: relative;
  padding-top: var(--pad2);
  padding-bottom: var(--pad2);
  flex-wrap: wrap;
}

.loadMore {
  text-align: center;
  padding: var(--pad4) 0;
}

:global {
  // We need global classes for react-transition-group.
  // https://reactcommunity.org/react-transition-group/transition-group
  .g-BookmarkList-item {
    &-enter {
      opacity: 0.01;
    }

    &-enter-active {
      opacity: 1;
      transition: opacity 80ms ease-in;
    }

    &-exit {
      opacity: 0.8;
    }

    &-exit-active {
      opacity: 0.01;
      transition: opacity 160ms linear;
    }
  }
}
