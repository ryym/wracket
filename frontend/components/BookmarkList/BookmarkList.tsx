import React from 'react';
import {Bookmark} from '../../lib/Bookmark';

export interface Props {
  readonly bookmarks: Bookmark[];
}

export function BookmarkList({bookmarks}: Props) {
  const items = bookmarks.map(b => (
    <li key={b.id}>
      <a href={b.url} target="_blank">
        {b.title}
      </a>
    </li>
  ));
  return (
    <div>
      <ul>{items}</ul>
    </div>
  );
}
