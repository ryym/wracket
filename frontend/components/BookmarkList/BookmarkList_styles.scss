@import '@material/icon-button/mixins';
@import '../../styles/vars/responsive';

.list {
  composes: mdc-list mdc-list--two-line from global;

  .listItem {
    // Override default height 72px of '.mdc-list--two-line .mdc-list-item'.
    height: auto;
  }
}

.listItem {
  composes: mdc-list-item from global;
  border-bottom: 1px solid var(--cl-primary-50);
  min-height: 72px;
  position: relative;
  padding-top: var(--pad2);
  flex-wrap: wrap;
}

.itemThumbnail {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-right: var(--pad4);
  flex: 0 0 92px;
  height: 92px;
  width: 92px;

  @include screen-sm {
    flex: 0 0 68px;
    height: 68px;
    width: 68px;
  }

  > img {
    border-radius: var(--pad);
    max-width: 100%;
    max-height: 100%;
  }
}

.itemBlankThumbnail {
  background-color: rgba(#d3e6ee, 0.3);
  border-radius: var(--pad);
  width: 88%;
  height: 88%;
}

.itemLink {
  color: var(--mdc-theme-text-primary-on-background);
  text-decoration: none;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  width: 100%;
}

.itemLinkText {
  composes: mdc-list-item__text from global;
  font-size: 1.6rem;
  margin: var(--pad2) 0;
}

.itemTitle {
  line-height: 1.4em;
  display: inline-block;
  padding: var(--pad2) 0;
  white-space: initial;
}

.itemDomain {
  composes: mdc-list-item__secondary-text from global;
  font-size: 1.4rem;
}

.actions {
  position: absolute;
  top: 0;
  right: 4px;

  // TODO: Consider beter UI on small devices.
  @include screen-sm {
    position: initial;
    width: 100%;
  }
}

.action {
  @include mdc-icon-button-size(18px, 18px, 4px);
  color: #b0bec5;

  &:hover {
    color: var(--cl-primary-900);
  }

  // Remove mdc-icon-button effects on ':active'.
  &::before,
  &::after {
    opacity: 0;
    background-color: transparent;
  }

  &.isFavorite {
    color: var(--cl-secondary-400);
  }
}
