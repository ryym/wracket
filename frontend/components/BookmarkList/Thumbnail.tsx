import * as React from 'react';

export type Props = {
  src: string;
} & {[key: string]: any};

type State = {
  src?: string;
};

// Thumbnail simply wraps img tag to lazy load the src.
export class Thumbnail extends React.PureComponent<Props, State> {
  state: State = {};

  componentDidMount() {
    setTimeout(() => this.setState({src: this.props.src}));
  }

  render() {
    const {state} = this;
    const {src, ...props} = this.props;
    if (state.src == null) {
      return null;
    }
    return <img src={state.src} {...props} />;
  }
}
