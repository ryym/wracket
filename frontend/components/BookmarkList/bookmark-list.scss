.bookmark-list_item {
  border-bottom: 1px solid #ccc;
  padding: var(--pad2);
}

.bookmark-list_actions {
  /* XXX: temporary */
  display: none;
}
