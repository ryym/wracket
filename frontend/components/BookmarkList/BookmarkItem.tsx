import * as React from 'react';
import {Bookmark, BookmarkStatus} from '../../lib/models';
import {IconButton} from '../form/IconButton';
import {Thumbnail} from './Thumbnail';

const cls = require('./BookmarkItem_styles.scss');

export type ItemProps = {
  readonly onBookmarkOpen?: (b: Bookmark) => void;
  readonly onBackToUnread?: (b: Bookmark) => void;
  readonly onFavoriteToggle?: (b: Bookmark, favorite: boolean) => void;
  readonly onArchiveClick?: (b: Bookmark) => void;
  readonly onReaddClick?: (b: Bookmark) => void;
  readonly onDeleteClick?: (b: Bookmark) => void;
};

export type Props = ItemProps & {
  readonly bookmark: Bookmark;
};

type State = {
  actionsOpen: boolean;
};

const noop = () => {};

export class BookmarkItem extends React.PureComponent<Props, State> {
  state = {
    actionsOpen: false,
  };

  toggleActions = () => {
    this.setState({actionsOpen: !this.state.actionsOpen});
  };

  render() {
    const {
      bookmark: b,
      onBookmarkOpen = noop,
      onBackToUnread = noop,
      onFavoriteToggle = noop,
      onArchiveClick = noop,
      onReaddClick = noop,
      onDeleteClick = noop,
    } = this.props;
    const {actionsOpen} = this.state;
    const isArchived = b.status === BookmarkStatus.Archived;
    return (
      <>
        <div className={cls.itemLinkContainer}>
          <a
            href={b.url}
            target="_blank"
            rel="nofollow noopener noreferrer"
            className={cls.itemLink}
            onClick={() => onBookmarkOpen(b)}
          >
            <div className={cls.itemThumbnail}>
              <Thumbnail src={b.thumbnailUrl} />
            </div>
            <div className={cls.itemLinkText}>
              <span className={cls.itemTitle}>{b.title || b.url}</span>
              <span className={cls.itemDomain}>{new URL(b.url).hostname}</span>
            </div>
          </a>
          <div className={cls.actionsOpener}>
            <IconButton
              className={cls.actionOpenerButton}
              content={actionsOpen ? 'expand_less' : 'expand_more'}
              label=""
              onClick={this.toggleActions}
            />
          </div>
        </div>
        <div className={`${cls.actions} ${actionsOpen ? cls.isOpen : ''}`}>
          {b.status === BookmarkStatus.Reading && (
            <IconButton
              className={cls.action}
              content="undo"
              label="Back to unread"
              onClick={() => onBackToUnread(b)}
            />
          )}
          {isArchived && (
            <IconButton
              className={cls.action}
              content="add"
              label="Readd bookmark"
              onClick={() => onReaddClick(b)}
            />
          )}
          {!isArchived && (
            <IconButton
              className={cls.action}
              content="done"
              label="Archive bookmark"
              onClick={() => onArchiveClick(b)}
            />
          )}
          <IconButton
            className={cls.action}
            content="delete"
            label="Delete bookmark"
            onClick={() => confirm('delete permanently?') && onDeleteClick(b)}
          />
          <IconButton
            className={`${cls.action} ${b.favorite ? cls.isFavorite : ''}`}
            content="star"
            label={b.favorite ? 'Remove bookmark from favrites' : 'Add bookmark to favorites'}
            onClick={() => {
              onFavoriteToggle(b, !b.favorite);
            }}
          />
        </div>
      </>
    );
  }
}
