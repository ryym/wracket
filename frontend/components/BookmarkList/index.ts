import './bookmark-list.scss';
export * from './BookmarkList';
