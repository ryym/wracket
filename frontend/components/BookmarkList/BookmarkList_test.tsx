import * as React from 'react';
import {shallow} from 'enzyme';
import {BookmarkList} from './BookmarkList';

describe('<BookmarkList>', () => {
  it('accepts empty bookmarks', () => {
    const wrapper = shallow(<BookmarkList bookmarks={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
