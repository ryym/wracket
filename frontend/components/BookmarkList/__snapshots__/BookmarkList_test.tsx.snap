// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookmarkList> accepts empty bookmarks 1`] = `
<div
  className=""
>
  <TransitionGroup
    childFactory={[Function]}
    className="list"
    component="ul"
  />
</div>
`;
